### ISSUE TITLE
[2.6 beta1 w/ dco] SIGUSR1 causing crash. persist-remote-ip but with static remote ip specified(?) 

### ISSUE THREAD

**Issue by Originalimoc** [OpenVPN/openvpn#187]
**Describe the bug**
with persist-tun:
```
2022-12-07 16:13:31 UDPv4 link local: (not bound)
2022-12-07 16:13:31 UDPv4 link remote: [AF_INET]**127.0.0.1:12345**
2022-12-07 16:13:31 dco_do_write: netlink reports error (-1): Unspecific failure
2022-12-07 16:13:31 dco_do_write: failed to send netlink message: No route to host (-113)
2022-12-07 16:13:31 write UDPv4 []: Success (fd=4,code=0)
2022-12-07 16:13:33 dco_do_write: netlink reports error (-1): Unspecific failure
2022-12-07 16:13:33 dco_do_write: failed to send netlink message: No route to host (-113)
2022-12-07 16:13:33 write UDPv4 []: Success (fd=4,code=0)
```
then stuck, not crashing, a SIGHUP can "fix" it

without persist-tun:
```
2022-12-07 16:13:31 UDPv4 link local: (not bound)
2022-12-07 16:13:31 UDPv4 link remote: [AF_INET]127.0.0.1:12345
```
crashed

**To Reproduce**
Ubuntu 22. dco on.

client config:
```
remote 127.0.0.1:12345
persist-local-ip
persist-remote-ip
persist-tun
persist-key
```

**Expected behavior**
2.5.8 good. remove persist-remote-ip is good as 2.5.8.

**Version information (please complete the following information):**
 - OS: Ubuntu 22.04
 - OpenVPN version: 2.6 beta1
 - Repeat for peer if relevant

**Comment by cron2** [comment]
That client config looks incomplete - please show the full client config (without the actual key material, of course).

Is this part of a real setup?  Talking to 127.0.0.1 would not work normally for a production setup.

**Comment by Originalimoc** [comment]
yes, 127.0.0.1:12345 is another tunnel, actual remote route manually bypassed

**Comment by cron2** [comment]
"yes" is not an answer to "please show full client config" - if I'm to fix this issue, I need to understand what you are trying to achieve.

**Comment by Originalimoc** [comment]
wait I'm copying config

**Comment by Originalimoc** [comment]
A very generic config

**Comment by cron2** [comment]
> A very generic config

there is lots of stuff in your config that makes little sense, like `fast-io` or `ping-restart 3600` together with `ping 0`.  Also `persist-local-ip` only makes sense in a config where you have `bind` and `local $domainname`... and `auth-nocache` makes no sense if you don't actually have any cacheable authentication enabled.

That said, it's still likely that the combination of `--persist-remote-ip` and DCO is broken, so we'll have a very close look.

**Comment by Originalimoc** [comment]
I don't want ping static interval(avoid some detection), and ping-restart 3600 can be ping-restart 360000 or whatever because it'll almost never get triggered. auth-nocache is for password auth I used before cert + TLS. No idea what fast-io does indeed, ovpn doesn't use sendmmsg(?) for some reason?

**Comment by cron2** [comment]
Both `ping` and `ping-restart` default to 0 = off.

`sendmmsg()` is not used because nobody coded it yet, but with DCO this point has become moot - data path can now go into the kernel, and userland packet handling can stay as it is (single packet at a time).  Initial experiments with `sendmmsg()` did not resulted in such a great improvement, but at the same time made the code more complex.

Another question, tho - what is listening on 127.0.0.1:12345?  Is this a local OpenVPN instance?

**Comment by Originalimoc** [comment]
Yes after DCO this doesn't matter anymore. 12345 is an obfuscation tunnel.

**Comment by ordex** [comment]
the localport can be specified with `--lport` and it will remain static.

**Comment by cron2** [comment]
> the localport can be specified with `--lport` and it will remain static.

as long as there is `nobind`, nobody cares about `lport`...

**Comment by Originalimoc** [comment]
LOL checked log it did change but it's so fast that I never noticed. Leave it as it is XD.

**Comment by ordex** [comment]
running my client this way, but can't generate any crash:
```
openvpn --dev tun 
--client 
--ca ../../test-pki/pki/ca.crt 
--cert ../../test-pki/pki/issued/client1.crt 
--key ../../test-pki/pki/private/client1.key 
--verb 3 
--persist-local-ip 
--persist-remote-ip
--persist-tun 
--persist-key 
--fast-io 
--connect-retry 1 3 
--nobind 
--allow-compression no 
--auth-nocache 
--script-security 2 
--ping 0 
--ping-restart 3600 
--replay-window 5000 3 
--mlock 
--tun-mtu 1428 
--remote 127.0.0.1 12345
```

maybe it is also related to what is listening on the other side?

**Comment by ordex** [comment]
I get
```
2022-12-07 11:43:34 UDPv4 link local: (not bound)
2022-12-07 11:43:34 UDPv4 link remote: [AF_INET]127.0.0.1:12345
2022-12-07 11:43:34 read UDPv4 [ECONNREFUSED]: Connection refused (fd=3,code=111)
```
but even if I add some dumb listener on that port, I get nothing
```
2022-12-07 11:49:38 UDPv4 link local: (not bound)
2022-12-07 11:49:38 UDPv4 link remote: [AF_INET]127.0.0.1:12345
```

**Comment by ordex** [comment]
oh ok, I missed that. thanks

**Comment by Originalimoc** [comment]
Huh it turns out it's a bad idea to use static local port.
Because server is not getting an explicit-exit-notify when client is getting a SIGUSR1, it will wait until TLS timeout if port is same.
Expected behavior or bug?

**Comment by cron2** [comment]
please do not mix unrelated problems in one GH issue.  Also, you talk about "client getting SIGUSR1" and show a log file from the server getting a SIGINT.  Please open a new issue about the --explicit-exit-notify problem, and include client and server logs, with --verb 3 each.

**Comment by Originalimoc** [comment]
I'd like to do so.
Later.

**Comment by ordex** [comment]
Can we close this ticket? I don't think there has been any other report about this issue? And this also was reported on an very old version at this point.
